File: src/backend/app/schemas/enums.py
"""SSOT-3 §5-3: Enum types for API request/response schemas.

Nine enum types used across all API endpoints.
These are *schema-layer* enums (str, Enum) — separate from SQLAlchemy model enums.
"""

from enum import Enum


class LifecycleStage(str, Enum):
    """17-value lifecycle stage (SSOT-2 §2-1 / SSOT-4 §3-9)."""

    DISCOVERED = "discovered"
    SCORED = "scored"
    UNDER_REVIEW = "under_review"
    PLANNED = "planned"
    SKIPPED = "skipped"
    READING_QUEUED = "reading_queued"
    READING_IN_PROGRESS = "reading_in_progress"
    READING_COMPLETED = "reading_completed"
    READING_FAILED = "reading_failed"
    JUDGING_QUEUED = "judging_queued"
    JUDGING_IN_PROGRESS = "judging_in_progress"
    JUDGING_COMPLETED = "judging_completed"
    JUDGING_FAILED = "judging_failed"
    CHECKLIST_GENERATING = "checklist_generating"
    CHECKLIST_ACTIVE = "checklist_active"
    CHECKLIST_COMPLETED = "checklist_completed"
    ARCHIVED = "archived"


class CaseStatus(str, Enum):
    """Case status (human-driven)."""

    NEW = "new"
    REVIEWED = "reviewed"
    PLANNED = "planned"
    SKIPPED = "skipped"
    ARCHIVED = "archived"


class Verdict(str, Enum):
    """Eligibility verdict."""

    ELIGIBLE = "eligible"
    INELIGIBLE = "ineligible"
    UNCERTAIN = "uncertain"


class ChecklistItemStatus(str, Enum):
    """Checklist item check status."""

    PENDING = "pending"
    DONE = "done"


class TriggeredBy(str, Enum):
    """Event trigger source."""

    SYSTEM = "system"
    USER = "user"
    BATCH = "batch"
    CASCADE = "cascade"


class IncludeParam(str, Enum):
    """GET /cases/:id include parameter values (§4-1)."""

    CARD_CURRENT = "card_current"
    ELIGIBILITY_CURRENT = "eligibility_current"
    CHECKLIST_CURRENT = "checklist_current"
    LATEST_EVENTS = "latest_events"


class SortField(str, Enum):
    """Allowed sort fields for GET /cases (§4-1)."""

    DEADLINE_AT = "deadline_at"
    SCORE = "score"
    FIRST_SEEN_AT = "first_seen_at"
    CASE_NAME = "case_name"
    NEEDS_REVIEW = "needs_review"


class SortDirection(str, Enum):
    """Sort direction."""

    ASC = "asc"
    DESC = "desc"


class RetryScope(str, Enum):
    """Retry scope for retry-* endpoints (§4-2)."""

    SOFT = "soft"
    FORCE = "force"

File: src/frontend/src/types/enums.ts
/**
 * Enum types — SSOT-2 §2-1 / SSOT-3 §5-3.
 *
 * Must stay in sync with backend app/schemas/enums.py.
 */

// ---------------------------------------------------------------------------
// Lifecycle stages (17 values — SSOT-2 §2-1)
// ---------------------------------------------------------------------------

export type LifecycleStage =
  | 'discovered'
  | 'scored'
  | 'under_review'
  | 'planned'
  | 'skipped'
  | 'reading_queued'
  | 'reading_in_progress'
  | 'reading_completed'
  | 'reading_failed'
  | 'judging_queued'
  | 'judging_in_progress'
  | 'judging_completed'
  | 'judging_failed'
  | 'checklist_generating'
  | 'checklist_active'
  | 'checklist_completed'
  | 'archived';

export const LIFECYCLE_STAGES: readonly LifecycleStage[] = [
  'discovered',
  'scored',
  'under_review',
  'planned',
  'skipped',
  'reading_queued',
  'reading_in_progress',
  'reading_completed',
  'reading_failed',
  'judging_queued',
  'judging_in_progress',
  'judging_completed',
  'judging_failed',
  'checklist_generating',
  'checklist_active',
  'checklist_completed',
  'archived',
] as const;

// ---------------------------------------------------------------------------
// Stage groups (for filtering)
// ---------------------------------------------------------------------------

export type StageGroup =
  | 'Discovery'
  | 'Reading'
  | 'Judging'
  | 'Preparation'
  | 'Archive';

export const STAGE_GROUPS: Record<StageGroup, LifecycleStage[]> = {
  Discovery: ['discovered', 'scored', 'under_review', 'planned', 'skipped'],
  Reading: [
    'reading_queued',
    'reading_in_progress',
    'reading_completed',
    'reading_failed',
  ],
  Judging: [
    'judging_queued',
    'judging_in_progress',
    'judging_completed',
    'judging_failed',
  ],
  Preparation: ['checklist_generating', 'checklist_active', 'checklist_completed'],
  Archive: ['archived'],
};

// ---------------------------------------------------------------------------
// UI label & color mapping (SSOT-2 §2-1)
// ---------------------------------------------------------------------------

export interface StageMeta {
  label: string;
  color: string;
  pulse?: boolean;
}

export const STAGE_META: Record<LifecycleStage, StageMeta> = {
  discovered: { label: '新着', color: 'gray' },
  scored: { label: 'スコア済', color: 'blue' },
  under_review: { label: '確認中', color: 'blue' },
  planned: { label: '応札予定', color: 'indigo' },
  skipped: { label: '見送り', color: 'gray' },
  reading_queued: { label: '読解待ち', color: 'yellow' },
  reading_in_progress: { label: '読解中…', color: 'yellow', pulse: true },
  reading_completed: { label: '読解完了', color: 'green' },
  reading_failed: { label: '読解失敗', color: 'red' },
  judging_queued: { label: '判定待ち', color: 'yellow' },
  judging_in_progress: { label: '判定中…', color: 'yellow', pulse: true },
  judging_completed: { label: '判定完了', color: 'green' },
  judging_failed: { label: '判定失敗', color: 'red' },
  checklist_generating: { label: '生成中…', color: 'yellow', pulse: true },
  checklist_active: { label: '準備中', color: 'orange' },
  checklist_completed: { label: '準備完了', color: 'green' },
  archived: { label: 'アーカイブ', color: 'gray' },
};

// ---------------------------------------------------------------------------
// Other enums
// ---------------------------------------------------------------------------

export type CaseStatus = 'new' | 'reviewed' | 'planned' | 'skipped' | 'archived';

export type Verdict = 'eligible' | 'ineligible' | 'uncertain';

export type ChecklistItemStatus = 'pending' | 'done';

export type TriggeredBy = 'system' | 'user' | 'batch' | 'cascade';

export type SortField =
  | 'deadline_at'
  | 'score'
  | 'first_seen_at'
  | 'case_name'
  | 'needs_review';

export type SortDirection = 'asc' | 'desc';

export type RiskLevel = 'low' | 'medium' | 'high';

export type ExtractionMethod = 'text' | 'ocr' | 'text_failed';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** True when the stage indicates processing is in progress. */
export function isProcessingStage(stage: LifecycleStage): boolean {
  return (
    stage.endsWith('_queued') ||
    stage.endsWith('_in_progress') ||
    stage === 'checklist_generating'
  );
}

/** True when the stage represents a stable state. */
export function isStableStage(stage: LifecycleStage): boolean {
  return !isProcessingStage(stage);
}

/** True when the stage represents a failure. */
export function isFailedStage(stage: LifecycleStage): boolean {
  return stage.endsWith('_failed');
}
